File: denngay/api/send-reminder.js
export default async function handler(req, res) {
  // Allow GET to be triggered by Vercel Cron, and POST for testing
  if (req.method !== 'GET' && req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  // Secure the endpoint if CRON_SECRET is set in Vercel environment variables
  if (process.env.CRON_SECRET) {
    const authHeader = req.headers.authorization;
    if (authHeader !== `Bearer ${process.env.CRON_SECRET}`) {
      return res.status(401).json({ error: 'Unauthorized' });
    }
  }

  const botToken = process.env.TELEGRAM_BOT_TOKEN;
  const chatId = process.env.TELEGRAM_CHAT_ID;

  if (!botToken || !chatId) {
    return res.status(500).json({ error: 'Thiếu cấu hình Telegram Token hoặc Chat ID' });
  }

  // Define interactive buttons
  const replyMarkup = {
    inline_keyboard: [
      [
        { text: "Đã uống 🌸", callback_data: "taken" },
        { text: "Để tí nữa ⏰", callback_data: "later" },
        { text: "Nay em nghỉ 💤", callback_data: "off" }
      ]
    ]
  };

  const url = `https://api.telegram.org/bot${botToken}/sendMessage`;
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        chat_id: chatId,
        text: "Đến giờ uống thuốc rồi em iu ơi! 🌸",
        reply_markup: replyMarkup
      })
    });

    if (!response.ok) {
      const errText = await response.text();
      return res.status(500).json({ error: `Telegram error: ${errText}` });
    }

    return res.status(200).json({ success: true });
  } catch (err) {
    return res.status(500).json({ error: err.message });
  }
}

File: send_reminder.py
import sys
import os
import urllib.request
import urllib.parse
import urllib.error
import json

# Cấu hình UTF-8 cho console để tránh lỗi hiển thị tiếng Việt trên Windows
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8')
if hasattr(sys.stderr, 'reconfigure'):
    sys.stderr.reconfigure(encoding='utf-8')

# Tải cấu hình từ file .env nếu có (khi chạy thử local)
env_file = os.path.join(os.path.dirname(__file__), '.env')
if os.path.exists(env_file):
    with open(env_file, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#'):
                key, val = line.split('=', 1)
                os.environ[key.strip()] = val.strip()

token = os.environ.get('TELEGRAM_BOT_TOKEN')
chat_id = os.environ.get('TELEGRAM_CHAT_ID')

if not token or not chat_id:
    print("Lỗi: Thiếu biến môi trường TELEGRAM_BOT_TOKEN hoặc TELEGRAM_CHAT_ID.")
    exit(1)

# Định nghĩa giao diện nút bấm tương tác
reply_markup = {
    "inline_keyboard": [
        [
            {"text": "Đã uống 🌸", "callback_data": "taken"},
            {"text": "Để tí nữa ⏰", "callback_data": "later"},
            {"text": "Nay em nghỉ 💤", "callback_data": "off"}
        ]
    ]
}

url = f"https://api.telegram.org/bot{token}/sendMessage"
data = urllib.parse.urlencode({
    'chat_id': chat_id,
    'text': "Đến giờ uống thuốc rồi em iu ơi! 🌸",
    'reply_markup': json.dumps(reply_markup)
}).encode('utf-8')

req = urllib.request.Request(url, data=data, method='POST')
req.add_header('Content-Type', 'application/x-www-form-urlencoded')

try:
    print("Đang gửi tin nhắc nhở kèm nút bấm tương tác...")
    with urllib.request.urlopen(req) as response:
        res_body = response.read().decode('utf-8')
        res_json = json.loads(res_body)
        if res_json.get("ok"):
            print("Tin nhắn nhắc nhở đã được gửi thành công!")
        else:
            print(f"Gửi tin nhắn thất bại: {res_body}")
except urllib.error.HTTPError as he:
    err_body = he.read().decode('utf-8')
    print(f"Đã xảy ra lỗi HTTP {he.code}: {err_body}")
    exit(1)
except Exception as e:
    print(f"Đã xảy ra lỗi hệ thống: {e}")
    exit(1)
